src/filter.ts

```diff
@@ -19,208 +34,283 @@ function match(pattern: any, value: any) {
+      valid(transformedValue)
+    )[1];
+    return predicate(transformedValue);
+  };
 }
 
 const FILTERS: Record<
   string,
-  (hass: HassObject, value: any, entity: HAState) => Promise<boolean>
+  (hass: HassObject, value: any) => Promise<(entity: HAState) => boolean>
 > = {
-  options: async () => true,
-  sort: async () => true,
-  domain: async (hass, value, entity) => {
-    return match(value, entity.entity_id.split(".")[0]);
+  options: async (hass, value) => (entity) => true,
+  sort: async (hass, value) => (entity) => true,
+  domain: async (hass, value) => {
+    const [matcher] = await Promise.all([match(value)]);
+    return (entity) => matcher(entity.entity_id.split(".")[0]);
   },
-  entity_id: async (hass, value, entity) => {
-    return match(value, entity.entity_id);
+  entity_id: async (hass, value) => {
+    const [matcher] = await Promise.all([match(value)]);
+    return (entity) => matcher(entity.entity_id);
   },
-  state: async (hass, value, entity) => {
-    return match(value, entity.state);
+  state: async (hass, value) => {
+    const [matcher] = await Promise.all([match(value)]);
+    return (entity) => matcher(entity.state);
   },
-  name: async (hass, value, entity) => {
-    return match(value, entity.attributes?.friendly_name);
+  name: async (hass, value) => {
+    const [matcher] = await Promise.all([match(value)]);
+    return (entity) => matcher(entity.attributes?.friendly_name);
   },
-  group: async (hass, value, entity) => {
+  group: async (hass, value) => (entity) => {
     return hass.states[value]?.attributes?.entity_id?.includes(
       entity.entity_id
     );
   },
-  attributes: async (hass, value, entity) => {
-    for (const [k, v] of Object.entries(value as Record<string, any>)) {
-      let attr = k.split(" ")[0]; // Remove any suffixes
-      let obj = entity.attributes;
-      for (const step of attr.split(":")) {
-        obj = obj ? obj[step] : undefined;
-      }
-      if (obj === undefined || !match(v, obj)) return false;
-    }
-    return true;
+  attributes: async (hass, value: Record<string, any>) => {
+    let filters = (
+      await Promise.all(
+        Object.entries(value).map(async ([k, v]) => {
+          const attr = k.split(" ")[0]; // Remove any suffixes
+          const steps = attr.split(":");
+          const stepper = (obj) => steps.reduce((a, x) => a?.[x], obj);
+          const matcher = await match(v);
+          return { stepper, matcher };
+        })
+      )
+    ).map(({ stepper, matcher }) => (entity) => {
+      const obj = stepper(entity.attributes);
+      return obj !== undefined && matcher(obj);
+    });
+    return (entity) => !filters.map((filter) => filter(entity)).includes(false);
   },
-  not: async (hass, value, entity) => {
-    return !(await filter_entity(hass, value, entity.entity_id));
+  not: async (hass, value) => {
+    const [filter] = await Promise.all([filter_entity(hass, value)]);
+    return (entity) => !filter(entity.entity_id);
```

Comment:
`return (entity) => !filter(entity)`

Reply:
Tried making this change on my branch today, and realized that `filter` is `(entity_id: string) => boolean`, so passing the `entity_id` is correct here.